File: src/trocado/features/identity/application/use_cases/sign_in_use_case.py
from __future__ import annotations

import asyncio

from trocado.core.application.interfaces.clock_interface import ClockInterface
from trocado.core.application.interfaces.identifier_provider_interface import (
    IdentifierProviderInterface,
)
from trocado.features.identity.application.data.session_data import SessionData
from trocado.features.identity.application.data.sign_in_data import SignInData
from trocado.features.identity.application.interfaces.password_hasher_interface import (
    PasswordHasherInterface,
)
from trocado.features.identity.application.interfaces.person_repository_interface import (
    PersonRepositoryInterface,
)
from trocado.features.identity.application.interfaces.session_repository_interface import (
    SessionRepositoryInterface,
)
from trocado.features.identity.application.interfaces.token_generator_interface import (
    TokenGeneratorInterface,
)
from trocado.features.identity.application.mappers.session_data_mapper import SessionDataMapper
from trocado.features.identity.domain.entities.session_entity import SessionEntity
from trocado.features.identity.domain.errors.invalid_credentials_error import InvalidCredentialsError
from trocado.features.identity.domain.errors.invalid_email_error import InvalidEmailError
from trocado.features.identity.domain.errors.weak_password_error import WeakPasswordError
from trocado.features.identity.domain.value_objects.email_value_object import EmailValueObject
from trocado.features.identity.domain.value_objects.password_value_object import PasswordValueObject
from trocado.features.identity.domain.virtual_objects.authenticated_session_virtual_object import (
    AuthenticatedSessionVirtualObject,
)

# A real Argon2 digest of a throwaway secret, used ONLY on the not-found path so that *every* sign-in pays
# exactly one verify. Without it, an unknown email would return faster than a wrong password and that timing
# gap would itself reveal whether the email exists. Its result is always discarded — it authenticates no one.
# Regenerate if the hasher's Argon2 parameters change materially (the cost must stay comparable to a real one).
_DECOY_HASH = "$argon2id$v=19$m=65536,t=3,p=4$k1hgbldg8ljpxdmJuLmHWg$gHdBhBY09Aev0XcDQ+Ezm2BfR6JhDwbX1AhKFSBZWp8"


class SignInUseCase:
    """Verify a credential and, on success, issue a session — the `sign_in` half of identity.

    Two security rules shape the verification, both serving the same end (no account enumeration):
      - **One generic failure.** A malformed email, an unknown/inactive email, and a wrong password all
        raise `InvalidCredentialsError` — the caller can never tell which half was wrong.
      - **Uniform timing.** A verify runs on *every* attempt; when no person is found, it runs against a
        constant decoy hash. This deliberately inverts the usual "cheap guard before the expensive call":
        here the hash cost is paid on purpose, so response time never betrays the email's existence.

    Only **past a successful verify** is a session issued: a fresh `SessionEntity` with an opaque CSPRNG
    token and a fixed expiry, persisted, then returned as `SessionData` (the token + expiry + the person).
    A failed sign-in issues nothing. The raw password lives only as a transient value object and is never
    persisted, logged, or echoed.
    """

    def __init__(
        self,
        clock: ClockInterface,
        hasher: PasswordHasherInterface,
        identifier: IdentifierProviderInterface,
        token_generator: TokenGeneratorInterface,
        person_repository: PersonRepositoryInterface,
        session_repository: SessionRepositoryInterface,
    ) -> None:
        self._clock = clock
        self._hasher = hasher
        self._identifier = identifier
        self._token_generator = token_generator
        self._person_repository = person_repository
        self._session_repository = session_repository

    async def execute(self, data: SignInData) -> SessionData:
        # Building the value objects validates the input. In sign-in a malformed email or a too-short
        # password must look exactly like a wrong credential — so collapse those into the generic error
        # instead of letting InvalidEmailError/WeakPasswordError leak which factor was malformed.
        try:
            email = EmailValueObject(data.email)
            password = PasswordValueObject(data.password)
        except (InvalidEmailError, WeakPasswordError) as error:
            raise InvalidCredentialsError() from error

        # Sequential: the verify depends on which person (if any) the email resolves to — no gather.
        person = await self._person_repository.find_active_by_email(email)

        # Always verify, even with no person, against the decoy — equalizing timing. Authenticate only when
        # a real person was found AND their hash verifies; the decoy branch can only ever fail.
        verified = await self._hasher.verify(password, person.password if person is not None else _DECOY_HASH)
        if person is None or not verified:
            raise InvalidCredentialsError()

        # Past the guard: issue the session. The three reads are mutually independent — gather them.
        created_at, id, token = await asyncio.gather(
            self._clock.now(),
            self._identifier.generate(),
            self._token_generator.generate(),
        )
        session = SessionEntity.create(id=id, token=token, person_id=person.id, created_at=created_at)
        await self._session_repository.create(session)

        return SessionDataMapper.to_data(AuthenticatedSessionVirtualObject(session=session, person=person))

File: src/main/kotlin/com/bed/cordato/features/identity/application/driving/use_cases/SignInUseCase.kt
package com.bed.cordato.features.identity.application.driving.use_cases

import com.bed.cordato.core.application.driven.ports.ClockPort
import com.bed.cordato.core.application.driven.ports.TokenizerPort
import com.bed.cordato.core.application.driven.ports.IdGeneratorPort
import com.bed.cordato.core.application.driven.repositories.SessionRepository
import com.bed.cordato.core.domain.entities.SessionEntity

import com.bed.cordato.features.identity.domain.errors.SignInError
import com.bed.cordato.features.identity.domain.value_objects.EmailValueObject

import com.bed.cordato.features.identity.application.driving.results.SignInResult
import com.bed.cordato.features.identity.application.driving.commands.SignInCommand
import com.bed.cordato.features.identity.application.driven.ports.PasswordHasherPort
import com.bed.cordato.features.identity.application.driven.repositories.PersonRepository

/**
 * Authenticates a person by e-mail and password and, on success, opens a session — the system's
 * first session producer.
 *
 * The order is the deliberate **inverse** of signup's "check existence before hashing" optimisation:
 * here `verify` is **always** paid, against the person's real hash when one exists and against a
 * fixed dummy hash of equivalent cost when it does not. Skipping the verify for an unknown e-mail
 * would turn response time into an account-discovery oracle — the very thing the single generic
 * [SignInError.InvalidCredentials] and the `401` hide from the body and status. Wrong password,
 * unknown e-mail, a malformed e-mail, and a non-active person therefore all collapse into the same
 * failure, indistinguishable by outcome or by timing.
 *
 * The public `invoke` signature is the driving (primary) side of this context.
 */
class SignInUseCase(
    private val clock: ClockPort,
    private val tokenizer: TokenizerPort,
    private val hasher: PasswordHasherPort,
    private val generator: IdGeneratorPort,
    private val personRepository: PersonRepository,
    private val sessionRepository: SessionRepository,
) {
    operator fun invoke(command: SignInCommand): SignInResult {
        val person = EmailValueObject.of(command.email)?.let(personRepository::findByEmail)

        // Always one verify, real or dummy — the cost must not reveal whether the e-mail exists.
        val authenticated = hasher.verify(command.password, person?.hash ?: DUMMY_HASH)
        if (person == null || !authenticated) {
            return SignInResult.Failure(SignInError.InvalidCredentials)
        }

        val now = clock()
        val token = tokenizer.generate()
        val session = SessionEntity(
            id = generator(),
            createdAt = now,
            personId = person.id,
            hashToken = tokenizer.hash(token),
            expiresAt = now.plus(SessionEntity.TTL),
        )
        // A false open means a token-hash collision (unreachable for a SecureRandom token); it is an
        // infrastructure anomaly, not a credentials refusal, so surface it as a 500 rather than lie 200.
        check(sessionRepository.open(session)) { "Failed to open a session for an authenticated person" }

        return SignInResult.Success(session, token)
    }

    private companion object {
        /**
         * A well-formed bcrypt hash the verify runs against when no real person is found, so an
         * unknown e-mail pays the same hashing cost as a wrong password. It MUST share the cost factor
         * of [com.bed.cordato.features.identity.infrastructure.adapters.PasswordHasherAdapter] (12) —
         * a cheaper dummy would reopen the timing oracle. No password ever matches it (that is fine;
         * only the cost matters).
         */
        const val DUMMY_HASH = "\$2a\$12\$BDGLFH8gboVw1dbjlZvOeeLcWmBcMa3bkpbeW5O3GGDPm5a5j/i.."
    }
}
